docker-app/qfieldcloud/core/tests/test_jobs.py

```diff
@@ -0,0 +1,173 @@
+
+        # "valid" localized layer -> QFC considers it as invalid
+        valid_localized_layer = processfile_layers[VALID_LOCALIZED_LAYER_KEY]
+        self.assertFalse(valid_localized_layer["is_valid"])
+        self.assertTrue(valid_localized_layer["is_localized"])
+
+        # invalid localized layer
+        invalid_localized_layer = processfile_layers[INVALID_LOCALIZED_LAYER_KEY]
+        self.assertFalse(invalid_localized_layer["is_valid"])
+        self.assertTrue(invalid_localized_layer["is_localized"])
+        self.assertEquals(
+            invalid_localized_layer["error_code"], "localized_dataprovider"
+        )
+
+        # invalid layer (datasource does not exist)
+        invalid_layer = processfile_layers[INVALID_LAYER_KEY]
+        self.assertFalse(invalid_layer["is_valid"])
+        self.assertFalse(invalid_layer["is_localized"])
+        self.assertEquals(invalid_layer["error_code"], "invalid_dataprovider")
+
+    def test_bad_layer_handler_values_for_package_job(self):
+        # Test that BadLayerHandler is parsing data properly during package jobs
+        self.client.credentials(HTTP_AUTHORIZATION="Token " + self.t1.key)
+
+        # Push the QGIS project and datasource GPKG files
+        response = self.upload_file(self.p1.id, "bumblebees.gpkg", "bumblebees.gpkg")
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        response = self.upload_file(
+            self.p1.id, "simple_bumblebees_wrong_localized.qgs", "project.qgs"
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        # Create Package job and wait for it to finish
+        set_subscription(
+            self.u1,
+            "keep_10",
+            storage_mb=10,
+            is_premium=True,
+            is_external_db_supported=True,
+        )
+        package_job = PackageJob.objects.create(
+            type=Job.Type.PROCESS_PROJECTFILE,
+            project=self.p1,
+            created_by=self.u1,
+        )
+        self.assertEqual(package_job.status, Job.Status.PENDING)
+        wait_for_project_ok_status(self.p1)
+
+        package_job.refresh_from_db()
+        self.assertEqual(package_job.status, Job.Status.FINISHED)
+        self.assertIsNotNone(package_job.feedback)
+
+        # extract layer data from job
+        for step in package_job.feedback["steps"]:
+            if step["id"] == "qgis_layers_data":
+                qgis_layers = step["returns"]["layers_by_id"]
+            if step["id"] == "qfield_layer_data":
+                qfield_layers = step["returns"]["layers_by_id"]
+
+        for layers in [qgis_layers, qfield_layers]:
+            # "valid" localized layer -> QFC considers it as invalid
+            valid_localized_layer = layers[VALID_LOCALIZED_LAYER_KEY]
+            self.assertFalse(valid_localized_layer["is_valid"])
+            self.assertTrue(valid_localized_layer["is_localized"])
```

Comment:
Is there a way to avoid the for loop? You better write the code twice, but be more explicit in the tests.